File: course-02242-examples/src/executables/java/dtu/compute/exec/Simple.java
package dtu.compute.exec;

/**
 * This class tries to only contain very simple programs.
 */
public class Simple {

    @Case
    public static void noop() {
        return;
    }

    @Case
    public static int zero() {
        return 0;
    }

    @Case
    public static int hundredAndTwo() {
        return 102;
    }

    @Case
    public static int identity(int a) {
        return a;
    }

    @Case
    public static int add(int a, int b) {
        return a + b;
    }

    @Case
    public static int min(int a, int b) {
        if (a <= b) return a;
        else return b;
    }

    @Case
    public static int factorial(int n) {
        int m = n;
        while (n-- > 0) {
            m *= n;
        }
        return m;
    }
}

File: project/Interpreter/Simple.py
import json

from Interpreter import InterpretFunction


dir = "course-02242-examples\decompiled"
file = "dtu\compute\exec\Simple"
assert (
    InterpretFunction(
        dir=dir,
        file=file,
        function="noop",
        stack=[],
        memory={},
    )
    == None
)
assert (
    InterpretFunction(
        dir=dir,
        file=file,
        function="zero",
        stack=[],
        memory={},
    )
    == 0
)
assert (
    InterpretFunction(
        dir=dir,
        file=file,
        function="hundredAndTwo",
        stack=[],
        memory={},
    )
    == 102
)
assert (
    InterpretFunction(
        memory={0: {"type": "integer", "value": 20}},
        dir=dir,
        file=file,
        function="identity",
        stack=[],
    )
    == 20
)
assert (
    InterpretFunction(
        memory={
            0: {"type": "integer", "value": 10},
            1: {"type": "integer", "value": 5},
        },
        dir=dir,
        file=file,
        function="add",
        stack=[],
    )
    == 15
)

assert (
    InterpretFunction(
        memory={
            0: {"type": "integer", "value": 10},
            1: {"type": "integer", "value": 5},
        },
        dir=dir,
        file=file,
        function="min",
        stack=[],
    )
    == 5
)

assert (
    InterpretFunction(
        memory={
            0: {"type": "integer", "value": 10},
            1: {"type": "integer", "value": 5},
        },
        dir=dir,
        file=file,
        function="div",
        stack=[],
    )
    == 2
)
